// textlint-plugin-asciidoc-loose/lib/AsciidocProcessor.js
// LICENSE : MIT
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _parseToAst = require("./parse-to-ast");

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var AsciidocProcessor = function () {
    function AsciidocProcessor(config) {
        _classCallCheck(this, AsciidocProcessor);

        this.config = config;
    }

    _createClass(AsciidocProcessor, [{
        key: "processor",
        value: function processor(ext) {
            return {
                preProcess: function preProcess(text, filePath) {
                    return (0, _parseToAst.parse)(text);
                },
                postProcess: function postProcess(messages, filePath) {
                    return {
                        messages: messages,
                        filePath: filePath ? filePath : "<asciidoc>"
                    };
                }
            };
        }
    }], [{
        key: "availableExtensions",
        value: function availableExtensions() {
            // http://asciidoctor.org/docs/asciidoc-recommended-practices/
            return [".asciidoc", ".adoc", ".asc"];
        }
    }]);

    return AsciidocProcessor;
}();

exports.default = AsciidocProcessor;
module.exports = exports['default'];
